### PR TITLE
Update veRL to 0.7.0

### PR DESCRIPTION
**Pull request by chenyushuo**
## Description

As the title says.

### GSM8k

In the figure below, `test_trainer` denotes training conducted using Fully Sharded Data Parallel (FSDP), while `test-megatron_trainer` refers to training performed with Megatron.  
The training curve obtained with Megatron exhibits a constant scaling discrepancy relative to the corresponding FSDP curve, due to the omission of scaling the output `actor/kl_loss` by the number of micro-batches (`n_micro_batch`). This issue has been addressed in the latest code version.  
All other curves remain consistent with those from previous experiments.

image

### Countdown

In the figure below, `test_trainer` corresponds to FSDP-based training, and `test-megatron_trainer` corresponds to Megatron-based training.  
All curves align closely with those observed in prior experiments.

image

#### FSDP

In the figure below, `guru-bench` represents the experimental results from before this pull request (PR), whereas `guru_math-bench` reflects the results obtained in this PR.  
Due to the introduction of scaling for the output `actor/kl_loss` in this PR, a constant multiplicative difference is observed compared to historical curves. All other metrics remain consistent with prior experimental results.

image

#### Megatron

In the figure below, `fixed-moe-is-megatron` denotes the experimental results from before this PR, and `fixed-moe-is-megatron-verl-0.7.0` corresponds to the results from this PR.  
As with the FSDP case, the scaling of the output `actor/kl_loss` introduced in this PR leads to a constant multiplicative offset relative to previous curves. Aside from this adjustment, all other curves are consistent with earlier experimental outcomes.

image

### Bug Fix for Entropy Calculation in Megatron

The image below illustrates GPU memory utilization: the yellow box corresponds to our current implementation, while the red box shows the memory usage after applying the veRL fix.  
The veRL fix incurs higher GPU memory consumption than our implementation—attributable to the use of `.clone()` versus a monkey patch applied directly to the Megatron function.  
Consequently, we have temporarily commented out the veRL fix pending further optimization.

image

## Checklist

Please check the following items before code is ready to be reviewed.

- [ ]  Code has passed all tests
- [x]  Docstrings have been added/updated in Google Style
- [x]  Documentation has been updated
- [x]  Code is ready for review

### CODE REVIEW

File: trinity/trainer/verl_trainer.py [hunk 1]
```diff
@@ -295,42 +285,98 @@ def init_workers(self):
...
-            self.resource_pool_to_cls[resource_pool]["ref"] = ref_policy_cls
+            self.resource_pool_to_cls[resource_pool][str(Role.RefPolicy)] = ref_policy_cls
+
         # create a reward model if reward_fn is None
-        if self.use_rm:
-            # we create a RM here
-            resource_pool = self.resource_pool_manager.get_resource_pool(Role.RewardModel)
-            rm_cls = RayClassWithInitArgs(
-                self.role_worker_mapping[Role.RewardModel], config=self.config.reward_model
+        # for legacy discriminative reward model, we create a reward model worker here
```
Comment (pan-x-c) [inline trinity/trainer/verl_trainer.py]: The reward model can be removed

File: trinity/common/verl_config.py [hunk 2]
```diff
@@ -559,6 +600,13 @@ def synchronize_config(self, config: Config) -> None:  # noqa: C901
         for field_name in config.algorithm.optimizer.__dataclass_fields__:
             field_value = getattr(config.algorithm.optimizer, field_name)
             if field_name == "optimizer_type":
+                if config.trainer.trainer_strategy.startswith("fsdp"):
+                    optim_map = {
+                        "adam": "AdamW",
+                        "adamw": "AdamW",
+                        "sgd": "SGD",
+                    }
+                    field_value = optim_map.get(field_value, field_value)
```
Comment (pan-x-c) [inline trinity/common/verl_config.py]: Raise an error here if not found

File: trinity/trainer/verl/fsdp_workers.py [hunk 3]
```diff
@@ -488,6 +640,24 @@ def _build_model_optimizer(  # noqa: C901
 
         return actor_module_fsdp, actor_optimizer, actor_lr_scheduler, actor_model_config
 
+    async def trainer_mode(self):  # TODO: check this
+        """Context switch hybridengine to trainer mode."""
+        # if self.config.rollout.free_cache_engine:
```
Comment (pan-x-c) [inline trinity/trainer/verl/fsdp_workers.py]: Can be removed

File: trinity/common/verl_config.py (hunk 2 above)
Comment (chenyushuo) [reply trinity/common/verl_config.py]: I do not recommend raising a `ValueError` here, as veRL currently permits users—under FSDP—to import the appropriate optimizer specified by `optim.optimizer` from the `optim.optimizer_impl` package (which defaults to `torch.optim`). This mechanism is specifically intended to correct case mismatches, since optimizer names in `torch.optim` are capitalized, whereas the current default value of `optim.optimizer` is entirely lowercase.

File: trinity/trainer/verl/fsdp_workers.py (hunk 3 above)
Comment (chenyushuo) [reply trinity/trainer/verl/fsdp_workers.py]: ok

File: trinity/trainer/verl_trainer.py (hunk 1 above)
Comment (chenyushuo) [reply trinity/trainer/verl_trainer.py]: ok
